server/src/server.ts

```diff
@@ -1,172 +1,628 @@
-    return languageService.formatOnTypeProvider.getIndentEdit(
-      params.position.line,
-      params.position.character,
-      params.ch,
-      params.options.tabSize,
-      params.options.insertSpaces,
-    );
+  connection.onDocumentOnTypeFormatting(async (params) => {
+    return await dispatch(params, {
+      async sas(languageService) {
+        return languageService.formatOnTypeProvider.getIndentEdit(
+          params.position.line,
+          params.position.character,
+          params.ch,
+          params.options.tabSize,
+          params.options.insertSpaces,
+        );
+      },
+    });
   });
 
-  connection.onSignatureHelp((params) => {
-    const languageService = getLanguageService(params.textDocument.uri);
-    completionProvider = languageService.completionProvider;
-    return completionProvider.getSignatureHelp(
-      params.position,
-      params.context?.activeSignatureHelp?.activeSignature,
-    );
+  connection.onSignatureHelp(async (params, token) => {
+    return await dispatch(params, {
+      async sas(languageService) {
+        return await languageService.completionProvider.getSignatureHelp(
+          params.position,
+          params.context?.activeSignatureHelp?.activeSignature,
+        );
+      },
+      async python(pyrightLanguageService) {
+        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-explicit-any
+        return (await pyrightLanguageService.onSignatureHelp(
+          params,
+          token,
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        )) as any;
+      },
+    });
   });
 
   connection.onDocumentFormatting((params) => {
+    syncIfDocChange(params.textDocument.uri);
     const languageService = getLanguageService(params.textDocument.uri);
     return languageService.formatter.format({
       tabWidth: params.options.tabSize,
       useTabs: params.options.insertSpaces === false,
     });
   });
 
-  documents.onDidChangeContent((event) => {
-    if (servicePool[event.document.uri]) {
-      documentPool[event.document.uri] = event.document;
-      return;
-    }
-    servicePool[event.document.uri] = new LanguageServiceProvider(
-      event.document,
+  connection.onDidOpenTextDocument(async (params) => {
+    const doc = TextDocument.create(
+      params.textDocument.uri,
+      "sas",
+      params.textDocument.version,
+      params.textDocument.text,
+    );
+    documentPool[doc.uri] = { document: doc, changed: false };
+    await _pyrightLanguageProvider.onDidOpenTextDocument(params);
+  });
+
+  connection.onDidCloseTextDocument(async (params) => {
+    const uri = params.textDocument.uri;
+    delete documentPool[uri];
+    await _pyrightLanguageProvider.onDidCloseTextDocument(params);
+  });
+
+  connection.onDidChangeTextDocument((params) => {
+    const uri = params.textDocument.uri;
+    const docInfo = documentPool[uri];
+    TextDocument.update(
```

Comment:
why update document by yourself? why not use the `documents.onDidChangeContent` event as before?

Reply:
Because the PyrightServer needs onDidOpenTextDocument() and onDidCloseTextDocument() to be handled.
If use documents.onDidChangeContent() as before, it requires to also use documents.onDidOpen(), documents.onDidClose() to handle open/close event, otherwise it doesn't work. However, it's parameters are different from what Pyright accepts, so can't easily pass it to Pyright.